nltk/tokenize/ssp.py

```diff
@@ -0,0 +1,142 @@
+
+    >>> SSP = SonoritySequencingPrinciple()
+    >>> SSP.tokenize('justification')
+    ['jus', 'ti', 'fi', 'ca', 'tion']
+    '''
+
+    def __init__(self, sonority_hierarchy=False):
+        '''
+        Sonority hierarchy should be provided in descending order.
+        If vowels are spread across multiple levels, they should be
+        passed assigned self.vowels var together.
+        '''
+
+        if not sonority_hierarchy:
+            sonority_hierarchy = [
+                'aeiouy',  # vowels
+                'lmnrw',  # nasals
+                'zvsf',  # fricatives
+                'bcdgtkpqxhj'  # stops
+            ]
+
+        self.vowels = sonority_hierarchy[0]
+        self.char_map = {}
+        for i, level in enumerate(sonority_hierarchy):
+            for c in level:
+                self.char_map[c] = len(sonority_hierarchy) - i
+
+    def assign_values(self, token):
+        '''
+        Assigns each character its value from the sonority hierarchy
+        '''
+        self.syllables_values = []
+        for c in token:
+            try:
+                self.syllables_values.append((c, self.char_map[c]))
+            except KeyError:
+                print("Warning: '{}' not defined in hierarchy".format(c))
+
+    def validate_syllables(self, syllabified):
+        '''
+        Ensures each syllable has at least one vowel.
+        If the following syllable doesn't have vowel,
+        add it to the current one.
+        '''
+        valid_syllables = []
+        front = ""
+        for i, syllable in enumerate(syllabified):
+
+            if not any(char in self.vowels for char in syllable):
+                if len(valid_syllables) == 0:
+                    front += syllable
+                else:
+                    valid_syllables = valid_syllables[
+                        :-1] + [valid_syllables[-1] + syllable]
+            else:
+                if len(valid_syllables) == 0:
+                    valid_syllables.append(front + syllable)
+                else:
+                    valid_syllables.append(syllable)
+
+        return valid_syllables
+
+    def tokenize(self, token):
+        '''
+        Apply the SSP to return a list of syllables.
+        '''
+        # if only one vowel return word
+        if sum(token.count(x) for x in self.vowels) <= 1:
+            return [token]
+
+        # assign values from hierarchy
+        self.assign_values(token)
+
+        syllable_list = []
+        syllable = ''
+        for i, tup in enumerate(self.syllables_values):
+            c, v = tup
+
+            if i == 0:  # if it's the first char, append automatically
+                syllable += c
+
+            else:
+                # main algorithm
+                if (i < len(self.syllables_values) - 1):
+
+                    # these cases trigger syllable break
+                    if v == self.syllables_values[i + 1][1] and \
+                            v == self.syllables_values[i - 1][1]:
```

Comment:
This is a good indication that looping through a trigrams of `self.syllables_values` would be good here.